src/control/lib/daos/api/attribute.go

```diff
@@ -0,0 +1,268 @@
+
+#include <daos_task.h>
+*/
+import "C"
+
+type attrType int
+
+const (
+	poolAttr attrType = iota
+	contAttr
+)
+
+func (at attrType) String() string {
+	switch at {
+	case poolAttr:
+		return "pool"
+	case contAttr:
+		return "container"
+	default:
+		return "unknown"
+	}
+}
+
+func listDaosAttributes(hdl C.daos_handle_t, at attrType) ([]string, error) {
+	var rc C.int
+	expectedSize, totalSize := C.size_t(0), C.size_t(0)
+
+	switch at {
+	case poolAttr:
+		rc = daos_pool_list_attr(hdl, nil, &totalSize, nil)
+	/*case contAttr:
+	rc = daos_cont_list_attr(hdl, nil, &totalSize, nil)*/
+	default:
+		return nil, errors.Wrapf(daos.InvalidInput, "unknown attr type %d", at)
+	}
+	if err := daosError(rc); err != nil {
+		return nil, errors.Wrapf(err, "failed to list %s attributes", at)
+	}
+
+	if totalSize < 1 {
+		return nil, nil
+	}
+
+	attrNames := []string{}
+	expectedSize = totalSize
+	cNamesBuf := C.malloc(totalSize)
+	defer C.free(cNamesBuf)
+
+	switch at {
+	case poolAttr:
+		rc = daos_pool_list_attr(hdl, (*C.char)(cNamesBuf), &totalSize, nil)
+	/*case contAttr:
+	rc = daos_cont_list_attr(hdl, (*C.char)(buf), &totalSize, nil)*/
+	default:
+		return nil, errors.Wrapf(daos.InvalidInput, "unknown attr type %d", at)
+	}
+	if err := daosError(rc); err != nil {
+		return nil, errors.Wrapf(err, "failed to list %s attributes", at)
+	}
+
+	if err := iterStringsBuf(cNamesBuf, expectedSize, func(name string) {
+		attrNames = append(attrNames, name)
+	}); err != nil {
+		return nil, err
+	}
+
+	return attrNames, nil
+}
+
+// getDaosAttributes fetches the values for the given list of attribute names.
+// Uses the bulk attribute fetch API to minimize roundtrips.
+func getDaosAttributes(hdl C.daos_handle_t, at attrType, reqAttrNames []string) (daos.AttributeList, error) {
+	if len(reqAttrNames) == 0 {
+		attrNameList, err := listDaosAttributes(hdl, at)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to list %s attributes", at)
+		}
+		reqAttrNames = attrNameList
+	}
+	numAttr := len(reqAttrNames)
+
+	if numAttr == 0 {
+		return nil, nil
+	}
+
+	// First, build a slice of C strings for the requested attribute names.
+	cAttrNames := make([]*C.char, numAttr)
+	for i, name := range reqAttrNames {
+		if name == "" {
+			return nil, errors.Wrapf(daos.InvalidInput, "empty %s attribute name at index %d", at, i)
+		}
+		cAttrNames[i] = C.CString(name)
+	}
+	defer func(nameSlice []*C.char) {
+		for _, name := range nameSlice {
+			freeString(name)
+		}
+	}(cAttrNames)
+
+	// Next, create a slice of C.size_t entries to hold the sizes of the values.
+	// We have to do this first in order to know the buffer sizes to allocate
+	// before fetching the actual values.
+	cAttrSizes := make([]C.size_t, numAttr)
+	var rc C.int
+	switch at {
+	case poolAttr:
+		rc = daos_pool_get_attr(hdl, C.int(numAttr), &cAttrNames[0], nil, &cAttrSizes[0], nil)
+	/*case contAttr:
+	rc = daos_cont_get_attr(hdl, C.int(numAttr), &attrNames[0], nil, &attrSizes[0], nil)*/
```

Comment:
Same question -- if these are being left as placeholders, it would be better to add a TODO with a ticket number.

Reply:
Next PR (https://github.com/daos-stack/daos/pull/15721) is coming in hot right behind this one. :) Trying to make them a little easier to review in chunks.

Reply:
Okay, just wanted to make sure it wasn't forgotten. Sounds okay to me.